src/MCRun.jl

```diff
@@ -290,48 +593,95 @@ function ptmc_run!(mc_states, move_strat, mc_params, pot, ensemble, results)
+            @inbounds mc_states = mc_cycle!(mc_states, move_strat, mc_params, pot, ensemble, n_steps, a, v, r)
+            #verbose way to save the highest and lowest energies
+
+            if save_ham == false
+                for i_traj = 1:mc_params.n_traj
+                    if mc_states[i_traj].en_tot < ebounds[1]
+                        ebounds[1] = mc_states[i_traj].en_tot
+                    end
+
+                    if mc_states[i_traj].en_tot > ebounds[2]
+                        ebounds[2] = mc_states[i_traj].en_tot
+                    end
+                end
+            end
+            #update stepsizes
+            if rem(i, mc_params.n_adjust) == 0
+                for i_traj = 1:mc_params.n_traj
+                    update_max_stepsize!(mc_states[i_traj], mc_params.n_adjust, a, v, r)
+                end 
+            end
         end
-    end
     #re-set counter variables to zero
-    for i_traj = 1:mc_params.n_traj
-        mc_states[i_traj].count_atom = [0, 0]
-        mc_states[i_traj].count_vol = [0, 0]
-        mc_states[i_traj].count_rot = [0, 0]
-        mc_states[i_traj].count_exc = [0, 0]
-    end 
+        for i_traj = 1:mc_params.n_traj
+            mc_states[i_traj].count_atom = [0, 0]
+            mc_states[i_traj].count_vol = [0, 0]
+            mc_states[i_traj].count_rot = [0, 0]
+            mc_states[i_traj].count_exc = [0, 0]
+        end
+        #initialise histogram for non-saving hamiltonian 
+        if save_ham == false
+            delta_en_hist = initialise_histograms!(mc_params,mc_states,results, full_ham=false,e_bounds=ebounds)
+        end
+
+        println("equilibration done")
+        
+
+        if save == true
+            save_states(mc_params,mc_states,0,save_dir)
+        end
+    end
 
-    println("equilibration done")
 
     #main MC loop
-    for i = 1:mc_params.mc_cycles
-        @inbounds mc_states = mc_cycle!(mc_states, move_strat, mc_params, pot, ensemble, n_steps, a, v, r) 
-        #sampling step
-        if rem(i, mc_params.mc_sample) == 0
-            for i_traj=1:mc_params.n_traj
-                push!(mc_states[i_traj].ham, mc_states[i_traj].en_tot) #to build up ham vector of sampled energies
+
+    if restart == false
+
+        for i = 1:mc_params.mc_cycles
+            if save_ham == false
+                ptmc_cycle!(mc_states,results,move_strat, mc_params, pot, ensemble ,n_steps ,a ,v ,r, save_ham, save, i,save_dir;delta_en_hist=delta_en_hist)
+            else
+                ptmc_cycle!(mc_states,results,move_strat, mc_params, pot, ensemble ,n_steps ,a ,v ,r, save_ham, save, i,save_dir)
             end
-        end 
-        #step adjustment
-        if rem(i, mc_params.n_adjust) == 0
-            for i_traj = 1:mc_params.n_traj
-                update_max_stepsize!(mc_states[i_traj], mc_params.n_adjust, a, v, r)
-            end 
+            
         end
-    end
 
+    else #if restarting
+
+        for i = restartindex:mc_params.mc_cycles
```

Comment:
index i is important for save functions typically, so we start from the restart_index